File: codes/HellDream/300.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        if(nums.length==0) return 0;
        int[] dp = new int[nums.length];
        int res = 1;
        for(int i=1;i<nums.length;i++){
            for(int j=0;j<i;j++){
                if(nums[j]<nums[i]){
                    dp[i] = Math.max(dp[i],1+dp[j]);
                }
            }
            res = Math.max(dp[i]+1, res);
        }
        return res;
    }
}

File: codes/Garnetwzy/300.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> tail(nums.size(), 0);
        int size = 0;
        for(int num: nums) {
            int i = 0;
            int j = size;
            while(i < j) { // find first val > num
                int mid = i + (j - i) / 2;
                if(tail[mid] < num) {
                    i = mid + 1;
                }else{
                    j = mid;
                }
            }
            tail[i] = num;
            if(i == size) {
                size++;
            }
        }
        return size;
    }
};